Bluetooth: Host: Re-subscribe only after encrypt change

jori-nordic opened the pull request:
Wait until encrypt change to trigger the re-subscription quirk. Otherwise it could fail with insufficient security.

Also gate it behind a kconfig to make apparent its dependencies:

If `add_subscriptions()` is executed in the first place, that means that the device was bonded and thus that it has to encrypt the link eventually.

`BT_GATT_AUTO_SEC_REQ` should take care of that.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
Co-authored-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

Fixes #52195 (partially)

jori-nordic committed:
Bluetooth: Host: Re-subscribe only after encrypt change

Wait until encrypt change to trigger the re-subscription quirk.
Otherwise it could fail with insufficient security.

Also gate it behind a kconfig to make apparent its dependencies:

If `add_subscriptions()` is executed in the first place, that means that
the device was bonded and thus that it has to encrypt the link eventually.

`BT_GATT_AUTO_SEC_REQ` should take care of that.

Also add a new flag to `bt_gatt_subscribe_params` to not send the same CCC
write multiple times in some cases.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>
Co-authored-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

hermabe commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -62,6 +62,16 @@ config BT_EATT_AUTO_CONNECT
+	help
+	  Quirk: upon re-establishing a bonded connection, assumes the remote
+	  forgot the CCC values and sets them again. This is only done after
+	  encryption.
```
Maybe add a reference to the `BT_GATT_SUBSCRIBE_FLAG_NO_RESUB` flag?

hermabe commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5498,17 +5498,14 @@ void bt_gatt_connected(struct bt_conn *conn)
-#if defined(CONFIG_BT_GATT_CLIENT)
-	add_subscriptions(conn);
-#if defined(CONFIG_BT_GATT_AUTO_UPDATE_MTU)
+#if defined(CONFIG_BT_GATT_AUTO_UPDATE_MTU) && defined(CONFIG_BT_GATT_CLIENT)
```
`BT_GATT_CLIENT` depends on `BT_GATT_CLIENT`, so no need for the `&& defined(CONFIG_BT_GATT_CLIENT)`.

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5531,6 +5530,10 @@ void bt_gatt_encrypt_change(struct bt_conn *conn)
 	data.conn = conn;
 	data.sec = BT_SECURITY_L1;
 
+#if defined(BT_GATT_AUTO_RESUBSCRIBE)
+	add_subscriptions(conn);
+#endif
```
Since `add_subscriptions` is a static function, we could remove the guard from `add_subscriptions` and use `IS_ENABLED` here; the linker should still optimize the function out.

Thalley replied:
```suggestion
#endif /* CONFIG_BT_GATT_AUTO_RESUBSCRIBE */
```

Thalley commented:
Looks like `add_subscriptions` may be guarded by something else as well

jori-nordic commented:
> Looks like `add_subscriptions` may be guarded by something else as well

I just reverted it to being ifdef'd out.

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5204,6 +5204,7 @@ void bt_gatt_cancel(struct bt_conn *conn, void *params)
 	}
 }
 
+#if defined(BT_GATT_AUTO_RESUBSCRIBE)
```
```suggestion
#if defined(CONFIG_BT_GATT_AUTO_RESUBSCRIBE)
```

jori-nordic replied:
yepp copy-paste error, thanks

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5531,6 +5530,10 @@ void bt_gatt_encrypt_change(struct bt_conn *conn)
 	data.conn = conn;
 	data.sec = BT_SECURITY_L1;
 
+#if defined(BT_GATT_AUTO_RESUBSCRIBE)
```
```suggestion
#if defined(CONFIG_BT_GATT_AUTO_RESUBSCRIBE)
```

Thalley commented:
@jori-nordic Looks like this is causing a segementation fault in MCC. Do you want me to take a look at it, or will you do that yourself?

jori-nordic commented:
> @jori-nordic Looks like this is causing a segementation fault in MCC. Do you want me to take a look at it, or will you do that yourself?

I'd appreciate if you could take a look at it, I'm not familiar with MCC. I can if you don't have the time though.

Thalley commented:
> > @jori-nordic Looks like this is causing a segementation fault in MCC. Do you want me to take a look at it, or will you do that yourself?
> 
> I'd appreciate if you could take a look at it, I'm not familiar with MCC. I can if you don't have the time though. edit: it's weird because I remember a green CI yesterday.

The CI was probably green yesterday because of the misconfigured guard, so it was never enabled ;) 

I'll take a look at it today

Thalley commented:
> > > @jori-nordic Looks like this is causing a segementation fault in MCC. Do you want me to take a look at it, or will you do that yourself?
> > 
> > 
> > I'd appreciate if you could take a look at it, I'm not familiar with MCC. I can if you don't have the time though. edit: it's weird because I remember a green CI yesterday.
> 
> The CI was probably green yesterday because of the misconfigured guard, so it was never enabled ;)
> 
> I'll take a look at it today

I have not yet found the root cause of the segmentation fault, but I think the fix by this PR is incorrect. 

Consider the following scenario: 

1. Zephyr subscribes to (unencrypted) characteristic C1
2. Zephyr subscribes to (encrypted) characteristic C2
3. 2. triggers the auto encryption update
4. On encryption complete, C2 subscription is reattempted
5. On encryption complete, the `add_subscriptions` is called in GATT, resubscribing to C1 and C2 (again!)

This wasn't an issue in the connected callback, but since some subscriptions may have been done (or in progress) at the time of the encryption complete (security changed callback) we will attempt to reencrypt.

jori-nordic commented:
Yes that's true. I initially wanted to add some state to keep track of which chars we have sent subscribe requests for, but decided against it. We're interested if you have an idea for a better design.

Re-subscribing shouldn't be an issue, right?

jori-nordic commented:
Yeah that's a fair point. At least now with the kconfig, applications with a too high number of subscriptions can disable it.
I was thinking of adding a (guarded) flag to [struct gatt_sub](https://github.com/jori-nordic/zephyr/blob/ab1f77770e64c2a0d9751136746e5dbf656492ce/subsys/bluetooth/host/gatt.c#L65) to keep track of whether we already sent a subscription request to that char since gatt was connected.
What do you think about that?

Thalley commented:
> 



> Yeah that's a fair point. At least now with the kconfig, applications with a too high number of subscriptions can disable it.
> 
> I was thinking of adding a (guarded) flag to [struct gatt_sub](https://github.com/jori-nordic/zephyr/blob/ab1f77770e64c2a0d9751136746e5dbf656492ce/subsys/bluetooth/host/gatt.c#L65) to keep track of whether we already sent a subscription request to that char since gatt was connected. What do you think about that?

That could fix that issue, yes, but then that's basically just a duplicate of `gatt_sub_find(conn)`. The flag could indeed just be added to `gatt_sub`, so it's not a huge thing to implement I think. If we indeed want to keep `add_subscriptions` then I think it works as a workaround. 

Btw, found the issue with MCC: Due to the massive amount of subscriptions in MCS/MCC, we subscribe to each characteristic one at a time. The implementation expects that a single call do `do_subscribe` will result in a single callback to `subscribe_mcs_char_func`. However, with this change in this PR, `subscribe_mcs_char_func` will be called twice on the first subscription due to the encryption change. That means that `do_subscribe` will then be called twice with the same discover parameter, since MCC uses the `CONFIG_BT_GATT_AUTO_DISCOVER_CCC` feature, which causes some weird issues in GATT i believe (I haven't been able to figure out exactly what is going wrong in GATT as I am basically just seeing 
```
==77288== Thread 8:
==77288== Jump to the invalid address stated on the next line
==77288==    at 0x0: ???
==77288==    by 0x80DAC22: att_handle_rsp (att.c:884)
==77288==    by 0x80DD5A5: att_handle_find_info_rsp (att.c:2452)
==77288==    by 0x80DE216: bt_att_recv (att.c:2840)
==77288==    by 0x80D7EA9: l2cap_chan_recv (l2cap.c:2592)
==77288==    by 0x80D805E: bt_l2cap_recv (l2cap.c:2626)
==77288==    by 0x80CEC47: bt_acl_recv (conn.c:359)
==77288==    by 0x80CED33: bt_conn_recv (conn.c:377)
==77288==    by 0x80BDB61: hci_acl (hci_core.c:521)
==77288==    by 0x80C2D25: bt_recv (hci_core.c:3498)
```

from Valgrind, but all the values in `att.c:884` looks fine, but it is definitely caused by the subscription callback being called twice when it is only expected once.

jori-nordic commented:
Ignore the last push I'm not done yet :)

Thalley commented:
> Ignore the last push I'm not done yet :)

Are you done now? :D

jori-nordic commented:
> > Ignore the last push I'm not done yet :)
> 
> Are you done now? :D

yes I wanted a green CI before pinging you guys

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1821,6 +1821,12 @@ struct bt_gatt_subscribe_params {
 #if defined(CONFIG_BT_EATT)
 	enum bt_att_chan_opt chan_opt;
 #endif /* CONFIG_BT_EATT */
+#if defined(CONFIG_BT_GATT_AUTO_RESUBSCRIBE)
+	/* Marks whether the subscription was sent on-air since the connection
+	 * was established.
+	 */
+	bool sent_sub;
+#endif	/* CONFIG_BT_GATT_AUTO_RESUBSCRIBE */
```
What happened to the idea of putting it in `struct gatt_sub`? 

I'm not crazy about adding an internal field to the parameter struct, but if we do it this way:

1. `sent_sub` should be prefix with `_` to indicate an internal field
2. We should have a reference to this value only being valid if `BT_GATT_SUBSCRIBE_FLAG_NO_RESUB` is not set. 

Actually, since we already have something like `BT_GATT_SUBSCRIBE_FLAG_WRITE_PENDING` in the GATT flags, should we add this `sent_sub` as a flag there instead?

jori-nordic replied:
`gatt_sub` wouldn't have worked I think, because it's per-device, not per-ccc so it breaks in some use-cases.

> Actually, since we already have something like `BT_GATT_SUBSCRIBE_FLAG_WRITE_PENDING` in the GATT flags, should we add this `sent_sub` as a flag there instead?

This sounds better, I'll try doing that yeah.

Thalley replied:
> `gatt_sub` wouldn't have worked I think, because it's per-device, not per-ccc so it breaks in some use-cases.

Can you give one such case? Is there a use case where we only want to subscribe to a subset of the characteristics that are subscribed without `BT_GATT_SUBSCRIBE_FLAG_NO_RESUB`? 

Since `add_subscriptions` are only called once (or intended to be called once), I have trouble seeing how a per-device solution won't work. 

As I was writing the above, I just realized that `security_changed` may be called multiple times (e.g. going from Level 0 -> Level 2, and then from Level 2 to Level 3). Is that properly handled?

jori-nordic replied:
I had it in mind but can't seem to remember now. But adding the flag as you propose is better anyway, because it doesn't use any more memory.

The two `security_changed` calls won't be a problem because if `add_subscriptions` is called once, then all the chars that have been re-sent over will have the flag set (and won't be sent again).

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1781,6 +1781,16 @@ enum {
 	 */
 	BT_GATT_SUBSCRIBE_FLAG_WRITE_PENDING,
 
+	/** @brief Sent flag
+	 *
+	 *  If set, indicates that a subscription request (CCC write) has
+	 *  already been sent in the active connection.
+	 *
+	 *  Used to avoid sending subscription requests multiple times when the
+	 *  @kconfig{CONFIG_BT_GATT_AUTO_RESUBSCRIBE} quirk is enabled.
+	 */
+	BT_GATT_SUBSCRIBE_FLAG_SENT,
```
Both this and the `BT_GATT_SUBSCRIBE_FLAG_WRITE_PENDING` are internally used only, right? Whereas the `BT_GATT_SUBSCRIBE_FLAG_VOLATILE` and `BT_GATT_SUBSCRIBE_FLAG_NO_RESUB` flags are set by the user. 

Can't help but to wonder if they should be separated into internal and public flags, but that may be outside the scope of this PR, but perhaps we can at least add a comment here?

jori-nordic replied:
Yes you're right. Is it ok if I don't add the comment now, but make the other PR tomorrow?

jori-nordic replied:
I added the commit to this PR. I'm a bit unsure about it, on one end just renaming it like I did doesn't force the distinction, but adding another `ATOMIC_DEFINE` internal_flags would probably eat up one more word in memory for each subscribe params struct.
What do you think is best?

Thalley replied:
Personally I would much rather add an internal atomic and use that (quite little) extra memory, than putting internally used values in the public API. 

I'd like to get input from @jhedberg and @alwa-nordic on this. 

If we create the internal atomic, then we could also, after this PR, start the process of moving the `BT_GATT_SUBSCRIBE_FLAG_WRITE_PENDING` flag too, as that has no business being public either IMO.

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5204,19 +5211,25 @@ void bt_gatt_cancel(struct bt_conn *conn, void *params)
 	/* Lookup existing subscriptions */
 	SYS_SLIST_FOR_EACH_CONTAINER(&sub->list, params, node) {
-		if (bt_addr_le_is_bonded(conn->id, &conn->le.dst) &&
+		if (!atomic_test_bit(params->flags,
+				     BT_GATT_SUBSCRIBE_FLAG_SENT) &&
```
The `BT_GATT_SUBSCRIBE_FLAG_SENT` is ever only tested for bonded devices, but is unconditionally set and cleared. Is that intentional?

jori-nordic replied:
yes, I figured adding another conditional when setting and clearing it wasn't necessary. I can add it of course, but I think the flag's comment is clear on how it should be used.

Thalley replied:
Yeah, it's not really a problem :)

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1777,9 +1777,21 @@ enum {
 	 *
 	 *  If set, indicates write operation is pending waiting remote end to
-	 *  respond.
+	 *  respond. Internal use only.
```
```suggestion
	 *  respond. 
	 * 
	 * @note Internal use only.
```

Thalley replied:
This should make it clearer in the documentation

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1777,9 +1777,21 @@ enum {
+	 *  respond. Internal use only.
 	 */
-	BT_GATT_SUBSCRIBE_FLAG_WRITE_PENDING,
+	BT_GATT_SUBSCRIBE_INTERNAL_FLAG_WRITE_PENDING,
```
Renaming this is a stable API change, and would probably need to follow the process in https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#stable. 

I would suggest to leave the name as is, and keep the comment about it being internally used too.

jori-nordic replied:
don't really thing it's an API change since it was not written anywhere that this was a public flag.
I still see your point that technically it is though.

Thalley replied:
> don't really thing it's an API change since it was not written anywhere that this was a public flag.

Generally I'd assume that everything in `include/zephyr` is public, and before this PR it wasn't specified that this was an internal flag :P

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1777,9 +1777,21 @@ enum {
+	 *  Used to avoid sending subscription requests multiple times when the
+	 *  @kconfig{CONFIG_BT_GATT_AUTO_RESUBSCRIBE} quirk is enabled.
+	 *
+	 *  Internal use only.
```
```suggestion
	 *  @note Internal use only.
```

jori-nordic committed:
Bluetooth: host: clarify which GATT subscribe flags are internal

Add a note to the GATT subscription flags' docstring to indicate what flags
are used internally and are not part of the API.

We might want to move them in their own struct member, something like
`internal_flags` in a subsequent change.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

Thalley reviewed (dismissed):
Is OK as it. We can consider moving the internal flags to an internal field at some point.

jhedberg commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -62,6 +62,19 @@ config BT_EATT_AUTO_CONNECT
 endif # BT_EATT
 
+config BT_GATT_AUTO_RESUBSCRIBE
+	bool "Automatic re-subscription to characteristics."
```
minor issue: remove the `.` at the end. No other Kconfig option seem to use formal sentence termination for the description this way
